sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -67,9 +67,13 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
   const bool IsKernel = CommandGroup->getType() == CG::KERNEL;
   {
     std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+#ifdef _WIN32
     while (!Lock.owns_lock()) {
       Lock.try_lock();
     }
```

Comment:
How this resulted in deadlock in Linux? `Lock` is constructed with `defer_lock`.

Reply:
The root cause is the same as for Windows. But in this case, I would say that use of try_lock causes deadlock in libpthread.
There are two cases how we can get a deadlock on Linux if we will use spinlock. I will provide the only one:
We have 3 threads:
1 thread - calls scheduler.cpp: `waitForEvent()` -> `lock_shared()`
1 thread - calls commands.cpp:482: `while (MEnqueueStatus == EnqueueResultT::SyclEnqueueBlocked)`
2 thread - calls scheduler.cpp: `addHostAccessor()` -> `try_lock()` 
3 thread - calls scheduler.cpp: `waitForEvent()` -> `lock_shared()`
2 thread - I don't have source code for pthread, so it changes something in `__pthread_rwlock_trywrlock()` which prevents to execute lock_shared() correctly for 3 thread. Looks like after that the 2 thread hangs in while() loop inside `__pthread_rwlock_trywrlock()`.
3 thread - execute `futex_abstimed_wait()`

Reply:
It looks pretty strange.

Reply:
I'm OK with understanding this one:
> Repeatedly looping on tryLock() also has the effect of waiting for the lock if it is not available, but without allowing proper OS-controlled blocking and waking of the waiting thread. Instead the waiting thread will continue to occupy a CPU, burning CPU time and power, while waiting for the lock to be released.

Though, if I understand it correctly there's smth funny with scheduling of threads.